tools/create-hybrid-image

```diff
@@ -0,0 +1,312 @@
+#!/bin/bash
+
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright 2019 Joyent, Inc.
+#
+
+if [[ -n "$TRACE" ]]; then
+	export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+	set -o xtrace
+fi
+
+set -euo pipefail
+IFS=$'\n\t'
+
+ISO=
+IMAGE_NAME=
+DESC=
+HOMEPAGE=
+OUTPUT_DIR=.
+PROTOTYPE=
+
+BUILD_DATE=$(date +%Y%m%d)
+
+SDC_VMMANIFEST=$(cd "$(dirname "$0")"; pwd)/sdc-vmmanifest
+
+VM_UUID=$(uuidgen)
+
+usage() {
+cat <<EOF
+
+Create a hybrid (kvm, bhyve) Debian image from a given ISO file.  This should be
+run from a SmartOS (joyent brand) zone that has been set up for this purpose.
+
+Usage:
+  $0 -i <ISO> -n <IMAGE_NAME> -d <DESC> -u <HOMEPAGE> -p <PROTOTYPE> \
+      [-o <OUTDIR>] [-- QEMUARGS]
+
+Example:
+  $0 -i binary.iso -n debian-10 \\
+      -d "Debian 10 (bionic) 64-bit image with just essential packages... " \\
+      -u https://docs.joyent.com/images/hvm/debian
+
+OPTIONS:
+  -i The ISO
+  -n The name of the image as it would appear in the manifest
+  -d The description of the image
+  -o Directory in which to place output.  Default is ".".
+  -p Prototype manifest file
+  -u The homepage URL for the image
+  -h Show this message
+
+EOF
+}
+
+while getopts "hi:n:d:o:p:u:" OPTION; do
+	case $OPTION in
+	h)
+		usage
+		exit 0
+		;;
+	i)
+		ISO=${OPTARG}
+		;;
+	n)
+		IMAGE_NAME=${OPTARG}
+		;;
+	d)
+		DESC=${OPTARG}
+		;;
+	o)
+		OUTPUT_DIR=${OPTARG}
+		;;
+	p)
+		PROTOTYPE=${OPTARG}
+		;;
+	u)
+		HOMEPAGE=${OPTARG}
+		;;
+	?)
+		usage 1>&2
+		exit 1
+		;;
+	esac
+done
+
+if [[ -z ${ISO} || -z ${IMAGE_NAME} || -z ${DESC} || -z ${HOMEPAGE} || \
+    -z ${PROTOTYPE} ]]; then
+	echo "FATAL: All of -i, -n, -d, -p, and -u are required." 1>&2
+	usage 1>&2
+	exit 1
+fi
+
+if [[ ! -f $PROTOTYPE ]]; then
+	echo "FATAL: Manifest prototype file '$PROTOTYPE' does not exist" 1>&2
+	exit 1
+fi
+
+shift $((OPTIND - 1))
+if (( $# == 0 )); then
+	# Set to a scalar value that will work for a test later.  bash is
+	# unhappy if QEMU_ARGS is declared as an empty array and it is used.
+	QEMU_ARGS=
+else
+	QEMU_ARGS=( "$@" )
+fi
+
+PIDFILE=/tmp/qemu-$IMAGE_NAME.pid
+
+function sanity_check
+{
+	sane=true
+
+	topds=$(zfs list -Ho name /data || true)
```

Comment:
Ought we hardcode where our delegated dataset is? Could the user pass in -z to specify on, and have it compute the mountpoint of that dataset instead?

Reply:
ok